AsyncDisplayKit/TextKit/ASTextKitFontSizeAdjuster.mm

```diff
@@ -179,23 +181,47 @@ - (CGFloat)scaleFactor
       if (numberOfLines > _attributes.maximumNumberOfLines) {
         
         for (NSUInteger index = scaleIndex; index < scaleFactors.count; index++) {
-          NSMutableAttributedString *entireAttributedString = [[NSMutableAttributedString alloc] initWithAttributedString:textStorage];
-          [[self class] adjustFontSizeForAttributeString:entireAttributedString withScaleFactor:[scaleFactors[index] floatValue]];
-          
+          // we are going to have to scale the string more. Reset it so we apply the scale factor to the full font size, not the scaled font size
+          // e.g., Say we scaled 90% of the original 16pt font so the longest word will fit. scaledString now has a font size of 14.4. If 14.4pt is
+          // still too large to fit in our max number of lines we will scale by the next factor, say 80%. We don't want to multiple 14.4 * .8, but
+          // the original 16pt * .8 to get a point size of 12.8
+          scaledString = [[NSMutableAttributedString alloc] initWithAttributedString:textStorage];
           
           // save away this scale factor. Even if we don't fit completely we should still scale down
           adjustedScale = [scaleFactors[index] floatValue];
+          [[self class] adjustFontSizeForAttributeString:scaledString withScaleFactor:adjustedScale];
+          
+          // adjust here so we start at the proper place in our scale array if we are too tall
+          scaleIndex++;
           
-          if ([self lineCountForString:entireAttributedString] <= _attributes.maximumNumberOfLines) {
+          if ([self lineCountForString:scaledString] <= _attributes.maximumNumberOfLines) {
             // we fit! we are done
             break;
           }
         }
+      }
+    }
+    
+    // check to see if the string's height is greater that the constrained height
+    CGSize stringSize = [scaledString boundingRectWithSize:CGSizeMake(_constrainedSize.width, CGFLOAT_MAX) options:NSStringDrawingUsesLineFragmentOrigin context:nil].size;
+    if (stringSize.height > _constrainedSize.height) {
+      for (NSUInteger index = scaleIndex; index < scaleFactors.count; index++) {
+        // we are going to have to scale the string more. Reset it so we apply the scale factor to the full font size, not the scaled font size
+        scaledString = [[NSMutableAttributedString alloc] initWithAttributedString:textStorage];
         
+        // save away this scale factor. Even if we don't fit completely we should still scale down
+        adjustedScale = [scaleFactors[index] floatValue];
+        [[self class] adjustFontSizeForAttributeString:scaledString withScaleFactor:adjustedScale];
+        
+        stringSize = [scaledString boundingRectWithSize:CGSizeMake(_constrainedSize.width, CGFLOAT_MAX) options:NSStringDrawingUsesLineFragmentOrigin context:nil].size;
+        if (stringSize.height <= _constrainedSize.height) {
+          // we fit! we are done
+          break;
+        }
       }
```

Comment:
Is there any way we can combine the constrainedSize check with the maximum line number check? Looks to be a lot of similar code. Just a question, as I can imagine it actually becomes more convoluted with that change.